azure/services/aso/aso.go

```diff
@@ -0,0 +1,165 @@
+		}
+		var readyErr error
+		if cond := conds[i]; cond.Status != metav1.ConditionTrue {
+			switch {
+			case cond.Reason == conditions.ReasonReconciling.Name:
+				readyErr = newOperationNotDoneError(existing)
+			default:
+				readyErr = fmt.Errorf("resource is not Ready: %s", conds[i].Message)
+			}
+
+			if readyErr != nil {
+				if conds[i].Severity == conditions.ConditionSeverityError {
+					return nil, azure.WithTerminalError(readyErr)
+				}
+				return nil, azure.WithTransientError(readyErr, 20*time.Second)
+			}
+		}
+	}
+
+	// Construct parameters using the resource spec and information from the existing resource, if there is one.
+	parameters, err := spec.Parameters(ctx, deepCopyOrNil(existing))
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get desired parameters for resource %s/%s (service: %s)", resourceNamespace, resourceName, serviceName)
+	}
+
+	parameters.SetName(resourceName)
+	parameters.SetNamespace(resourceNamespace)
+
+	diff := cmp.Diff(existing, parameters)
+	if diff == "" {
+		log.V(2).Info("resource up to date")
+		return existing, nil
+	}
+
+	// Create or update the resource with the desired parameters.
+	logMessageVerbPrefix := "creat"
+	if existing != nil {
+		logMessageVerbPrefix = "updat"
+	}
+	log.V(2).Info(logMessageVerbPrefix+"ing resource", "diff", diff)
+	if existing != nil {
+		err = s.Client.Update(ctx, parameters)
+	} else {
+		err = s.Client.Create(ctx, parameters)
+	}
+	if err == nil {
+		// Resources need to be requeued to wait for the create or update to finish.
+		return nil, azure.WithTransientError(newOperationNotDoneError(parameters), 20*time.Second)
+	}
+	return nil, errors.Wrapf(err, fmt.Sprintf("failed to %se resource %s/%s (service: %s)", logMessageVerbPrefix, resourceNamespace, resourceName, serviceName))
+}
+
+// DeleteResource implements the logic for deleting a resource Asynchronously.
+func (s *Service) DeleteResource(ctx context.Context, spec azure.ASOResourceSpecGetter, serviceName string) (err error) {
+	ctx, log, done := tele.StartSpanWithLogger(ctx, "aso.Service.DeleteResource")
+	defer done()
+
+	resource := spec.ResourceRef()
+	resourceName := resource.GetName()
+	resourceNamespace := resource.GetNamespace()
+
+	log = log.WithValues("service", serviceName, "resource", resourceName, "namespace", resourceNamespace)
+
+	log.V(2).Info("deleting resource")
+	err = s.Client.Delete(ctx, resource)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// already deleted
+			log.V(2).Info("successfully deleted resource")
+			return nil
+		}
+		return errors.Wrapf(err, "failed to delete resource %s/%s (service: %s)", resourceNamespace, resourceName, serviceName)
+	}
+
+	return azure.WithTransientError(newOperationNotDoneError(resource), 20*time.Second)
+}
+
+func newOperationNotDoneError(resource metav1.Object) error {
+	return azure.NewOperationNotDoneError(&infrav1.Future{
+		Type:          "ASOReconcile",
```

Comment:
should this differentiate between Create/Update/Delete?

Reply:
We could, at least between create/update and delete. I'll add that.

AFAICT this value is only logged but doesn't affect any behavior.